server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,648 @@
+                IOUtils.close(in);
+            }
+        }
+    }
+
+    private static class TrackingReadBytesIndexInput extends IndexInput {
+        final IndexInput in;
+        final BytesReadTracker bytesReadTracker;
+        final long fileOffset;
+
+        TrackingReadBytesIndexInput(IndexInput in, long fileOffset, BytesReadTracker bytesReadTracker) {
+            super(in.toString());
+            this.in = in;
+            this.fileOffset = fileOffset;
+            this.bytesReadTracker = bytesReadTracker;
+        }
+
+        @Override
+        public void close() throws IOException {
+            in.close();
+        }
+
+        @Override
+        public long getFilePointer() {
+            return in.getFilePointer();
+        }
+
+        @Override
+        public void seek(long pos) throws IOException {
+            in.seek(pos);
+        }
+
+        @Override
+        public long length() {
+            return in.length();
+        }
+
+        @Override
+        public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
+            final IndexInput slice = in.slice(sliceDescription, offset, length);
+            return new TrackingReadBytesIndexInput(slice, fileOffset + offset, bytesReadTracker.createSliceTracker(offset));
+        }
+
+        @Override
+        public IndexInput clone() {
+            return new TrackingReadBytesIndexInput(in.clone(), fileOffset, bytesReadTracker);
+        }
+
+        @Override
+        public byte readByte() throws IOException {
+            bytesReadTracker.trackPositions(fileOffset + getFilePointer(), 1);
+            return in.readByte();
+        }
+
+        @Override
+        public void readBytes(byte[] b, int offset, int len) throws IOException {
+            bytesReadTracker.trackPositions(fileOffset + getFilePointer(), len);
+            in.readBytes(b, offset, len);
+        }
+    }
+
+    /**
+     * Lucene Codec organizes data field by field for doc values, points, postings, and norms; and document by document
+     * for stored fields and term vectors. BytesReadTracker then can simply track the min and max read positions.
+     * This would allow us to traverse only two ends of each partition.
+     */
+    private static class BytesReadTracker {
+        private long minPosition = Long.MAX_VALUE; // inclusive
+        private long maxPosition = Long.MIN_VALUE; // exclusive
+
+        BytesReadTracker createSliceTracker(long offset) {
+            return this;
+        }
+
+        void trackPositions(long position, int length) {
+            minPosition = Math.min(minPosition, position);
+            maxPosition = Math.max(maxPosition, position + length -1);
+        }
+
+
+        void resetBytesRead() {
+            minPosition = Long.MAX_VALUE;
+            maxPosition = Long.MIN_VALUE;
+        }
+
+        long getBytesRead() {
+            if (minPosition <= maxPosition) {
+                return maxPosition - minPosition;
```

Comment:
Hmm, should it be `maxPosition - minPosition + 1`? E.g. if a single byte is read, we would have minPosition == maxPosition, so getBytesRead would return 0 instead of 1?

Reply:
🤦